From 205fcf1b32e633a228795939f9670d23576cec20 Mon Sep 17 00:00:00 2001
From: Ahzo <Ahzo@tutanota.com>
Date: Sat, 19 Sep 2020 22:11:44 +0200
Subject: [PATCH 2/2] resolver: support breaks and conflicts on versioned
 provides

Previously, provides were not considered for versioned breaks/conflicts.

Versioned provides are allowed since policy version 4.4.0.

Closes: #866974, #867006
---
 src/generic/apt/aptitude_resolver_universe.cc | 6 ++----
 src/generic/apt/aptitude_resolver_universe.h  | 3 ---
 2 files changed, 2 insertions(+), 7 deletions(-)

diff --git a/src/generic/apt/aptitude_resolver_universe.cc b/src/generic/apt/aptitude_resolver_universe.cc
index 7ff50f0d..05477c81 100644
--- a/src/generic/apt/aptitude_resolver_universe.cc
+++ b/src/generic/apt/aptitude_resolver_universe.cc
@@ -222,10 +222,8 @@ inline void aptitude_resolver_version::dep_iterator::advance()
 	move_to_next_dep = false;
     }
   // If we weren't trying to iterate over a Provides list *and* the
-  // current dep is a non-versioned Conflicts, start such an
-  // iteration.
-  else if(!prv_open && is_conflict(dep->Type) &&
-	  !dep.TargetVer())
+  // current dep is a Conflicts, start such an iteration.
+  else if(!prv_open && is_conflict(dep->Type))
     {
       prv = dep.TargetPkg().ProvidesList();
       if(!prv.end()) // otherwise we should advance to the next dep.
diff --git a/src/generic/apt/aptitude_resolver_universe.h b/src/generic/apt/aptitude_resolver_universe.h
index 526f4828..485a6ea2 100644
--- a/src/generic/apt/aptitude_resolver_universe.h
+++ b/src/generic/apt/aptitude_resolver_universe.h
@@ -1051,9 +1051,6 @@ bool aptitude_resolver_dep::broken_under(const InstallationType &I) const
 	  // if(start_iter.IsIgnorable(prv_iter) == true)
 	  //   return false;
 
-	  if(start_iter.TargetVer() != NULL)
-	    return false;
-
 	  return I.version_of(aptitude_resolver_package(prv_iter.OwnerPkg(), cache)).get_ver() == prv_iter.OwnerVer();
 	}
     }
-- 
2.28.0

